Upcoming Icinga 2.11: Build error on FreeBSD (boost related)

## Describe the bug
The upcoming Icinga 2.11 fails to build on FreeBSD:

```
[1/82] /usr/bin/c++  -DBOOST_COROUTINES_NO_DEPRECATION_WARNING -DBOOST_FILESYSTEM_NO_DEPRECATED -Ithird-party/nlohmann_json -Ithird-party/utf8cpp/source -I. -Ilib -O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include -Qunused-arguments -fcolor-diagnostics -pthread -Winvalid-pch -std=c++0x -O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include -MD -MT plugins/CMakeFiles/check_nscp_api.dir/check_nscp_api.cpp.o -MF plugins/CMakeFiles/check_nscp_api.dir/check_nscp_api.cpp.o.d -o plugins/CMakeFiles/check_nscp_api.dir/check_nscp_api.cpp.o -c plugins/check_nscp_api.cpp          
FAILED: plugins/CMakeFiles/check_nscp_api.dir/check_nscp_api.cpp.o
/usr/bin/c++  -DBOOST_COROUTINES_NO_DEPRECATION_WARNING -DBOOST_FILESYSTEM_NO_DEPRECATED -Ithird-party/nlohmann_json -Ithird-party/utf8cpp/source -I. -Ilib -O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include -Qunused-arguments -fcolor-diagnostics -pthread -Winvalid-pch -std=c++0x -O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include -MD -MT plugins/CMakeFiles/check_nscp_api.dir/check_nscp_api.cpp.o
-MF plugins/CMakeFiles/check_nscp_api.dir/check_nscp_api.cpp.o.d -o plugins/CMakeFiles/check_nscp_api.dir/check_nscp_api.cpp.o -c plugins/check_nscp_api.cpp                 
In file included from plugins/check_nscp_api.cpp:14:
In file included from lib/base/io-engine.hpp:12:
In file included from /usr/local/include/boost/asio/deadline_timer.hpp:24:
In file included from /usr/local/include/boost/asio/basic_deadline_timer.hpp:26:
/usr/local/include/boost/asio/detail/io_object_impl.hpp:87:40: error: no member named 'get_executor' in 'boost::asio::execution_context'                                     
      implementation_executor_(context.get_executor(),
                               ~~~~~~~ ^
/usr/local/include/boost/asio/ip/basic_resolver.hpp:118:7: note: in instantiation of function template specialization 'boost::asio::detail::io_object_impl<boost::asio::detail::resolver_service<boost::asio::ip::tcp>, boost::asio::executor>::io_object_impl<boost::asio::execution_context>' requested here                                             
    : impl_(context)
      ^
lib/base/tcpsocket.hpp:40:16: note: in instantiation of function template specialization 'boost::asio::ip::basic_resolver<boost::asio::ip::tcp, boost::asio::executor>::basic_resolver<boost::asio::execution_context>' requested here
        tcp::resolver resolver (socket.get_executor().context());
                      ^
plugins/check_nscp_api.cpp:189:11: note: in instantiation of function template specialization 'icinga::Connect<boost::asio::basic_socket<boost::asio::ip::tcp, boost::asio::executor> >' requested here
                icinga::Connect(stream->lowest_layer(), host, port);
                        ^
1 error generated.
ninja: build stopped: subcommand failed.
*** Error code 1

```

The commit tag I used is 00ea2d5.

FreeBSD 12.0-RELEASE-p4 FreeBSD 12.0-RELEASE-p4 GENERIC  amd64

```
pkg info -x boost
boost-libs-1.70.0_2
```

--- a/lib/base/tcpsocket.hpp
+++ b/lib/base/tcpsocket.hpp
@@ -4,6 +4,7 @@
 #define TCPSOCKET_H
 
 #include "base/i2-base.hpp"
+#include "base/io-engine.hpp"
 #include "base/socket.hpp"
 #include <boost/asio/ip/tcp.hpp>
 #include <boost/asio/spawn.hpp>
@@ -37,7 +38,7 @@ void Connect(Socket& socket, const String& node, const String& service)
 {
 	using boost::asio::ip::tcp;
 
-	tcp::resolver resolver (socket.get_executor().context());
+	tcp::resolver resolver (IoEngine::Get().GetIoService());
 	tcp::resolver::query query (node, service);
 	auto result (resolver.resolve(query));
 	auto current (result.begin());
@@ -66,7 +67,7 @@ void Connect(Socket& socket, const String& node, const String& service, boost::a
 {
 	using boost::asio::ip::tcp;
 
-	tcp::resolver resolver (socket.get_executor().context());
+	tcp::resolver resolver (IoEngine::Get().GetIoService());
 	tcp::resolver::query query (node, service);
 	auto result (resolver.async_resolve(query, yc));
 	auto current (result.begin());

--- a/lib/remote/apilistener.cpp
+++ b/lib/remote/apilistener.cpp
@@ -417,7 +417,7 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const std
 {
 	namespace asio = boost::asio;
 
-	auto& io (server->get_executor().context());
+	auto& io (IoEngine::Get().GetIoService());
 
 	for (;;) {
 		try {
@@ -651,7 +651,7 @@ void ApiListener::NewClientHandlerInternal(boost::asio::yield_context yc, const 
 
 			endpoint->AddClient(aclient);
 
-			asio::spawn(client->get_executor().context(), [this, aclient, endpoint, needSync](asio::yield_context yc) {
+			asio::spawn(IoEngine::Get().GetIoService(), [this, aclient, endpoint, needSync](asio::yield_context yc) {
 				CpuBoundWork syncClient (yc);
 
 				SyncClient(aclient, endpoint, needSync);

--- a/lib/remote/httpserverconnection.cpp
+++ b/lib/remote/httpserverconnection.cpp
@@ -12,6 +12,7 @@
 #include "base/configtype.hpp"
 #include "base/defer.hpp"
 #include "base/exception.hpp"
+#include "base/io-engine.hpp"
 #include "base/logger.hpp"
 #include "base/objectlock.hpp"
 #include "base/timer.hpp"
@@ -20,6 +21,7 @@
 #include <limits>
 #include <memory>
 #include <stdexcept>
+#include <boost/asio/io_service.hpp>
 #include <boost/asio/spawn.hpp>
 #include <boost/beast/core.hpp>
 #include <boost/beast/http.hpp>
@@ -31,8 +33,13 @@ using namespace icinga;
 auto const l_ServerHeader ("Icinga/" + Application::GetAppVersion());
 
 HttpServerConnection::HttpServerConnection(const String& identity, bool authenticated, const std::shared_ptr<AsioTlsStream>& stream)
-	: m_Stream(stream), m_Seen(Utility::GetTime()), m_IoStrand(stream->get_executor().context()), m_ShuttingDown(false), m_HasStartedStreaming(false),
-	m_CheckLivenessTimer(stream->get_executor().context())
+	: HttpServerConnection(identity, authenticated, stream, IoEngine::Get().GetIoService())
+{
+}
+
+HttpServerConnection::HttpServerConnection(const String& identity, bool authenticated, const std::shared_ptr<AsioTlsStream>& stream, boost::asio::io_service& io)
+	: m_Stream(stream), m_Seen(Utility::GetTime()), m_IoStrand(io), m_ShuttingDown(false), m_HasStartedStreaming(false),
+	m_CheckLivenessTimer(io)
 {
 	if (authenticated) {
 		m_ApiUser = ApiUser::GetByClientCN(identity);

--- a/lib/remote/httpserverconnection.hpp
+++ b/lib/remote/httpserverconnection.hpp
@@ -8,6 +8,7 @@
 #include "base/tlsstream.hpp"
 #include <memory>
 #include <boost/asio/deadline_timer.hpp>
+#include <boost/asio/io_service.hpp>
 #include <boost/asio/io_service_strand.hpp>
 #include <boost/asio/spawn.hpp>
 
@@ -41,6 +42,8 @@ private:
 	bool m_ShuttingDown;
 	bool m_HasStartedStreaming;
 	boost::asio::deadline_timer m_CheckLivenessTimer;
+
+	HttpServerConnection(const String& identity, bool authenticated, const std::shared_ptr<AsioTlsStream>& stream, boost::asio::io_service& io);
 
 	void ProcessMessages(boost::asio::yield_context yc);
 	void CheckLiveness(boost::asio::yield_context yc);

--- a/lib/remote/jsonrpcconnection.cpp
+++ b/lib/remote/jsonrpcconnection.cpp
@@ -16,6 +16,7 @@
 #include "base/tlsstream.hpp"
 #include <memory>
 #include <utility>
+#include <boost/asio/io_service.hpp>
 #include <boost/asio/spawn.hpp>
 #include <boost/date_time/posix_time/posix_time_duration.hpp>
 #include <boost/system/system_error.hpp>
@@ -30,10 +31,16 @@ static RingBuffer l_TaskStats (15 * 60);
 
 JsonRpcConnection::JsonRpcConnection(const String& identity, bool authenticated,
 	const std::shared_ptr<AsioTlsStream>& stream, ConnectionRole role)
-	: m_Identity(identity), m_Authenticated(authenticated), m_Stream(stream),
-	m_Role(role), m_Timestamp(Utility::GetTime()), m_Seen(Utility::GetTime()), m_NextHeartbeat(0), m_IoStrand(stream->get_executor().context()),
-	m_OutgoingMessagesQueued(stream->get_executor().context()), m_WriterDone(stream->get_executor().context()), m_ShuttingDown(false),
-	m_CheckLivenessTimer(stream->get_executor().context()), m_HeartbeatTimer(stream->get_executor().context())
+	: JsonRpcConnection(identity, authenticated, stream, role, IoEngine::Get().GetIoService())
+{
+}
+
+JsonRpcConnection::JsonRpcConnection(const String& identity, bool authenticated,
+	const std::shared_ptr<AsioTlsStream>& stream, ConnectionRole role, boost::asio::io_service& io)
+	: m_Identity(identity), m_Authenticated(authenticated), m_Stream(stream), m_Role(role),
+	m_Timestamp(Utility::GetTime()), m_Seen(Utility::GetTime()), m_NextHeartbeat(0), m_IoStrand(io),
+	m_OutgoingMessagesQueued(io), m_WriterDone(io), m_ShuttingDown(false),
+	m_CheckLivenessTimer(io), m_HeartbeatTimer(io)
 {
 	if (authenticated)
 		m_Endpoint = Endpoint::GetByName(identity);

--- a/lib/remote/jsonrpcconnection.hpp
+++ b/lib/remote/jsonrpcconnection.hpp
@@ -11,7 +11,7 @@
 #include "base/workqueue.hpp"
 #include <memory>
 #include <vector>
-#include <boost/asio/deadline_timer.hpp>
+#include <boost/asio/io_service.hpp>
 #include <boost/asio/io_service_strand.hpp>
 #include <boost/asio/spawn.hpp>
 
@@ -79,6 +79,8 @@ private:
 	AsioConditionVariable m_WriterDone;
 	bool m_ShuttingDown;
 	boost::asio::deadline_timer m_CheckLivenessTimer, m_HeartbeatTimer;
+
+	JsonRpcConnection(const String& identity, bool authenticated, const std::shared_ptr<AsioTlsStream>& stream, ConnectionRole role, boost::asio::io_service& io);
 
 	void HandleIncomingMessages(boost::asio::yield_context yc);
 	void WriteOutgoingMessages(boost::asio::yield_context yc);

